server/internal/lsp/server/Diagnostics.go

```diff
@@ -0,0 +1,116 @@
+package server
+
+import (
+	"bytes"
+	"log"
+	"os/exec"
+	"strconv"
+	"strings"
+
+	"github.com/pherrymason/c3-lsp/internal/lsp/project_state"
+	"github.com/pherrymason/c3-lsp/pkg/cast"
+	"github.com/tliron/glsp"
+	protocol "github.com/tliron/glsp/protocol_3_16"
+)
+
+func (s *Server) RunDiagnostics(state *project_state.ProjectState, notify glsp.NotifyFunc, delay bool) {
+	if state.IsCalculatingDiagnostics() {
+		return
+	}
+
+	state.SetCalculateDiagnostics(true)
+
+	binary := "c3c"
+	if s.options.C3CPath.IsSome() {
+		binary = s.options.C3CPath.Get()
+	}
+	command := exec.Command(binary, "build", "--test")
+	command.Dir = state.GetProjectRootURI()
+
+	// set var to get the output
+	var out bytes.Buffer
+	var stdErr bytes.Buffer
+
+	// set the output to our variable
+	command.Stdout = &out
+	command.Stderr = &stdErr
+
+	runDiagnostics := func() {
+		err := command.Run()
+		log.Println("output:", out.String())
+		log.Println("output:", stdErr.String())
+		if err != nil {
+			log.Println("An error:", err)
+			errorInfo := extractErrors(stdErr.String())
+
+			diagnostics := []protocol.Diagnostic{
+				errorInfo.Diagnostic,
+			}
+
+			go notify(
+				protocol.ServerTextDocumentPublishDiagnostics,
+				protocol.PublishDiagnosticsParams{
+					URI:         errorInfo.File,
+					Diagnostics: diagnostics,
+				})
+		}
+
+		state.SetCalculateDiagnostics(false)
+	}
+
+	if delay {
+		s.diagnosticDebounced(runDiagnostics)
```

Comment:
@nikpivkin you seem to have some more experience in Go than me.
I'm unable to correctly debounce the diagnostic calculation: it should not run it until 2 seconds after last document change.

Server (`s` in this line) is storing an instance of `diagnosticDebounced` function so it does not get recreated every document, so in theory, calling again `s.diagnosticDebounced(runDiagnostics)` cancels previous run if not yet executed.

One can easily recreate this bug by spamming keystrokes in a document, and after 2 seconds of the first key press the diagnostics will run.